### PR TITLE
gcc: gcc13 -> gcc14 (again)

### PR DESCRIPTION
**Pull request by emilazy**
Here we go again…

The stable release has too many AArch64 issues, but they seem to have been fixed upstream; Arch is shipping a Git snapshot. The next stable release should be out by the time 25.05 is finishing up, so this should be okay as a temporary solution.

This is the GCC 14-20241116 snapshot. We use that identifier as our package version, diverging from our usual unstable version convention, because it identifies itself as “gcc (GCC) 14.2.1 20241116” and comes in `gcc-14-20241116.tar.xz`; 20241116 is therefore a useful version identifier to use verbatim, and 14.2.0-unstable-2024-11-16 would potentially be confusing for a version that calls itself 14.2.1. The next stable release from the GCC 14 branch will be 14.3.0, so there should be no ambiguity here.

Getting this version is a little complicated; we need the precompiled `flex(1)` output that these Git snapshots don’t include. Thankfully, the source file hasn’t changed since 14.2.0, so we can simply download 14.2.0 too and extract the precompiled file. A little merging finesse is required for the `aarch64-darwin` patch, which we also upgrade.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [x] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by emilazy** [comment]
https://github.com/NixOS/nixpkgs/pull/356545 turned out to be too hard, so I think this is the best way forward.

File: pkgs/development/compilers/gcc/patches/default.nix [hunk 1]
```diff
@@ -138,13 +138,23 @@ in
...
-    # There are no upstream release tags in https://github.com/iains/gcc-14-branch.
-    # 04696df09633baf97cdbbdd6e9929b9d472161d3 is the commit from https://github.com/gcc-mirror/gcc/releases/tag/releases%2Fgcc-14.2.0
-    name = "gcc-14-darwin-aarch64-support.patch";
-    url = "https://github.com/iains/gcc-14-branch/compare/04696df09633baf97cdbbdd6e9929b9d472161d3..gcc-14.2-darwin-r0.diff";
-    hash = "sha256-GEUz7KdGzd2WJ0gjX3Uddq2y9bWKdZpT3E9uZ09qLs4=";
-  }) ];
+  "14" = [
+    (fetchpatch {
+      name = "gcc-14-darwin-aarch64-support.patch";
+      url = "https://raw.githubusercontent.com/Homebrew/formula-patches/refs/heads/master/gcc/gcc-14.2.0-r2.diff";
```
[the remark this answers is not in the document]
Comment (emilazy) [reply pkgs/development/compilers/gcc/patches/default.nix]: Whoops, yes, I got this from a pinned URL but GitHub unpinned it when I went to the raw file.

**Review by fabianhjr — APPROVED** [review]

**Comment by FliegendeWurst** [comment]
I am getting a build failure on libsForQt5.kcoreaddons, but it looks like the error is actually in the c++ implementation provided by gcc:
```
In file included from /nix/store/m6dp9bprckv06z6lha0q8glzi8gbfrzz-gcc-14-20241116/include/c++/14-20241116/string_view:904,
                 from /nix/store/m6dp9bprckv06z6lha0q8glzi8gbfrzz-gcc-14-20241116/include/c++/14-20241116/bits/basic_string.h:47,
                 from /nix/store/m6dp9bprckv06z6lha0q8glzi8gbfrzz-gcc-14-20241116/include/c++/14-20241116/string:54,
                 from /nix/store/xs5bzsmd96jkjqb4h6fjrvbzwaagx1ik-qtbase-5.15.15-dev/include/QtCore/qbytearray.h:52,
                 from /nix/store/xs5bzsmd96jkjqb4h6fjrvbzwaagx1ik-qtbase-5.15.15-dev/include/QtCore/qstring.h:50,
                 from /nix/store/xs5bzsmd96jkjqb4h6fjrvbzwaagx1ik-qtbase-5.15.15-dev/include/QtCore/qobject.h:47,
                 from /nix/store/xs5bzsmd96jkjqb4h6fjrvbzwaagx1ik-qtbase-5.15.15-dev/include/QtCore/QObject:1,
                 from /build/kcoreaddons-5.116.0/src/lib/jobs/kjob.h:14,
                 from /build/kcoreaddons-5.116.0/src/lib/jobs/kcompositejob.h:13,
                 from /build/kcoreaddons-5.116.0/src/lib/jobs/kcompositejob.cpp:9:
/nix/store/m6dp9bprckv06z6lha0q8glzi8gbfrzz-gcc-14-20241116/include/c++/14-20241116/bits/string_view.tcc: In member function 'constexpr std::basic_string_view<_CharT, _Traits>::size_type std::basic_string_view<_CharT, _Traits>::rfind(_CharT, size>
/nix/store/m6dp9bprckv06z6lha0q8glzi8gbfrzz-gcc-14-20241116/include/c++/14-20241116/bits/string_view.tcc:119:7: error: '__glibcxx_enable_shared_from_this' was not declared in this scope
  119 |       size_type __size = this->_M_len;
      |       ^~~~~~~~~
/nix/store/m6dp9bprckv06z6lha0q8glzi8gbfrzz-gcc-14-20241116/include/c++/14-20241116/bits/string_view.tcc:120:11: error: '__size' was not declared in this scope; did you mean 'size'?
  120 |       if (__size > 0)
      |           ^~~~~~
      |           size
```
Do you agree?

**Comment by emilazy** [comment]
I am not sure, but I suspect the error is likely an unhelpful one pointing to a problem in the KDE code; that file [hasn’t changed substantially in years](https://gcc.gnu.org/git/?p=gcc.git;a=history;f=libstdc%2B%2B-v3/include/bits/string_view.tcc;h=a9ba668d05cfe5b13e8834c5be573434020208cf;hb=HEAD). It could be [this commit](https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=63a598deb0c9fcb8704af6ef90f0b905942b0276) but that seems to only relate to warnings, not errors.

**Comment by FliegendeWurst** [comment]
I've found another regression. The default boost version (1.81) does not compile with GCC 14 when using `cstdfloat_limits.hpp`.

See https://github.com/boostorg/math/issues/992, it is fixed in 1.85 (https://github.com/boostorg/math/commit/b3a21bcbcbd95c23da594ff53c3a48e38f0ff9f3).

**Comment by emilazy** [comment]
Perhaps it is time to bump our default Boost version, then? Though we should also probably backport the patch.

**Comment by emilazy** [comment]
https://github.com/NixOS/nixpkgs/pull/356102

**Comment by trofi** [comment]
`gltron` build regressed. Proposed the fix as:
- https://github.com/NixOS/nixpkgs/pull/367678

**Comment by trofi** [comment]
`lv` build regressed. Proposed the fix as:
- https://github.com/NixOS/nixpkgs/pull/367689

**Comment by trofi** [comment]
`gnome2.ORBit2` build regressed. Proposed the fix as:
- https://github.com/NixOS/nixpkgs/pull/368851

**Comment by trofi** [comment]
`gnome-recipes` build regressed. Proposed the fix as:
- https://github.com/NixOS/nixpkgs/pull/373297

**Comment by trofi** [comment]
`spaceFM` build regressed. Proposed the fix as:
- https://github.com/NixOS/nixpkgs/pull/373304

**Pull request merged**